pkg/instrumentation/flags.go

```diff
@@ -21,135 +21,213 @@ import (
+		*s = Sampling(v)
+	default:
+		return instrumentationError("invalid Sampling value of type %T: %v", obj, obj)
+	}
+	return nil
+}
+
+func (s *Sampling) Parse(value string) error {
+	switch strings.ToLower(value) {
+	case "disabled":
+		*s = Disabled
+	case "testing":
+		*s = Testing
+	case "production":
+		*s = Production
 	default:
-		return instrumentationError("invalid TraceConfig value of type %T: %v", obj, obj)
+		f, err := strconv.ParseFloat(value, 64)
+		if err != nil {
+			return instrumentationError("invalid Sampling value '%s': %v", value, err)
+		}
+		*s = Sampling(f)
 	}
 	return nil
 }
 
-// Sampler returns a trace.Sampler corresponding to the TraceConfig value.
-func (tc TraceConfig) String() string {
+// Sampler returns a trace.Sampler corresponding to/approximating the Sampling value.
+func (s Sampling) String() string {
 	switch {
-	case tc <= 0.005:
+	case s <= 0.005:
 		return "disabled"
-	case tc <= 0.1:
+	case s <= 0.1:
 		return "production"
-	case tc >= 0.95:
+	case s >= 0.95:
 		return "testing"
 	}
-	return strconv.FormatFloat(float64(tc), 'f', -1, 64)
+	return strconv.FormatFloat(float64(s), 'f', -1, 64)
 }
 
-// Sampler returns a trace.Sampler corresponding to the TraceConfig value.
-func (tc TraceConfig) Sampler() trace.Sampler {
+// Sampler returns a trace.Sampler corresponding to the Sampling value.
+func (s Sampling) Sampler() trace.Sampler {
 	switch {
-	case tc >= 0.95:
+	case s >= 0.95:
 		return trace.AlwaysSample()
-	case tc <= 0.005:
+	case s <= 0.005:
 		return trace.NeverSample()
 	default:
-		return trace.ProbabilitySampler(float64(tc))
+		return trace.ProbabilitySampler(float64(s))
 	}
 }
 
+func getenv(name, defval string, parsefn func(string) error) error {
+	if val := os.Getenv(name); val != "" {
+		err := parsefn(val)
+		if err == nil {
+			return nil
+		}
+		log.Error("invalid environment '%s'='%s': %v, using default '%s'", name, val, err, defval)
+	}
+	return parsefn(defval)
+}
+
 // defaultOptions returns a new options instance, all initialized to defaults.
 func defaultOptions() interface{} {
-	collector := os.Getenv("JAEGER_COLLECTOR")
-	agent := os.Getenv("JAEGER_AGENT")
-	metrics := os.Getenv("PROMETHEUS_ENDPOINT")
+	o := &options{}
 
-	if collector == "" {
-		collector = defaultCollector
-	}
-	if agent == "" {
-		agent = defaultAgent
+	type param struct {
+		defval  string
+		parsefn func(string) error
 	}
-	if metrics == "" {
-		metrics = defaultMetrics
+
+	params := map[string]param{
+		"JAEGER_COLLECTOR": {
+			defaultCollector,
+			func(v string) error { o.Collector = v; return nil },
+		},
+		"JAEGER_AGENT": {
+			defaultAgent,
+			func(v string) error { o.Agent = v; return nil },
+		},
+		"HTTP_ENDPOINT": {
+			defaultHTTP,
+			func(v string) error { o.HTTP = v; return nil },
+		},
+		"PROMETHEUS_EXPORTER": {
+			defaultPrometheus,
+			func(v string) error {
+				switch strings.ToLower(v) {
+				case "1", "on", "true", "enable", "yes":
+					o.Prometheus = true
+				}
```

Comment:
You either set Prometheus = false in the config, or you set the PROMETHEUS_EXPORTED env. var to 'no', for instance.

Reply:
Ah, sorry I mis-read the code 😄 

But, I think we should check the negative values (0, off, false etc), too, and return an error if we are not able to parse it. So that if the user specifies `PROMETHEUS_EXPORTER=foo` we will use the default, instead of blindly false.

Reply:
Ah, this was another conversation I forgot to re-open. Shouldn't we check for the negative values, too? Maybe cherry-pick the `ParseEnabled` function from #168 ?

Reply:
I was planning to rebase and switch to `ParseEnabled`.